File: app/src/main/java/com/app/walletcards/network/CardApiService.kt
package com.app.walletcards.network

import android.util.Log
import com.app.walletcards.model.ApplyCardRequest
import com.app.walletcards.model.ApplyCardResponse
import com.app.walletcards.model.CardDetailsResponse
import com.app.walletcards.model.CardResponse
import com.app.walletcards.model.SubUser
import com.app.walletcards.model.ThreeDSResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

object CardApiService {
    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    suspend fun subuseradd(subUser: SubUser): ApplyCardResponse? {
        return withContext(Dispatchers.IO) {
            try {
                val jsonBody = json.encodeToString(subUser)
                Log.d("CardApiService", "subuseradd request: $jsonBody")

                val requestBody =
                    jsonBody.toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url(ApiConfig.BASE_URL + "subuseradd")
                    .post(requestBody)
                    .addHeader("publickey", ApiConfig.PUBLIC_KEY)
                    .addHeader("secretkey", ApiConfig.SECRET_KEY)
                    .addHeader("Content-Type", "application/json")
                    .build()

                val response = client.newCall(request).execute()
                val responseBody = response.body?.string()
                Log.d("CardApiService", "subuseradd response: $responseBody")

                responseBody?.let {
                    json.decodeFromString(ApplyCardResponse.serializer(), it)
                }

            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    suspend fun applyForNewVirtualCard(applyCardRequest: ApplyCardRequest): ApplyCardResponse? {
        return withContext(Dispatchers.IO) {
            try {
                val jsonBody = json.encodeToString(applyCardRequest)
                Log.d("CardApiService", "applyForNewsubuserCard request: $jsonBody")

                val requestBody =
                    jsonBody.toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url(ApiConfig.BASE_URL + "digitalnewsubusercard")
                    .post(requestBody)
                    .addHeader("publickey", ApiConfig.PUBLIC_KEY)
                    .addHeader("secretkey", ApiConfig.SECRET_KEY)
                    .addHeader("Content-Type", "application/json")
                    .build()

                val response = client.newCall(request).execute()
                val responseBody = response.body?.string()
                Log.d("CardApiService", "applyForNewsubuserCard response: $responseBody")

                responseBody?.let {
                    json.decodeFromString(ApplyCardResponse.serializer(), it)
                }

            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    suspend fun getAllDigitalCards(userEmail: String): CardResponse? {
        return withContext(Dispatchers.IO) {
            try {
                val jsonBody = """
                    {
                        "useremail": "$userEmail"
                    }
                """.trimIndent()

                val requestBody =
                    jsonBody.toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url(ApiConfig.BASE_URL + "getsubuseralldigital")
                    .post(requestBody)
                    .addHeader("publickey", ApiConfig.PUBLIC_KEY)
                    .addHeader("secretkey", ApiConfig.SECRET_KEY)
                    .addHeader("Content-Type", "application/json")
                    .build()

                val response = client.newCall(request).execute()

                response.body?.string()?.let {
                    json.decodeFromString(CardResponse.serializer(), it)
                }

            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    suspend fun getDigitalCardDetails(email: String, cardId: String): CardDetailsResponse? {
        return withContext(Dispatchers.IO) {
            try {
                val jsonBody = """
                    {
                        "useremail": "$email",
                        "cardid": "$cardId"
                    }
                """.trimIndent()

                val requestBody =
                    jsonBody.toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url(ApiConfig.BASE_URL + "getsubuserdigitalcard")
                    .post(requestBody)
                    .addHeader("publickey", ApiConfig.PUBLIC_KEY)
                    .addHeader("secretkey", ApiConfig.SECRET_KEY)
                    .addHeader("Content-Type", "application/json")
                    .build()

                val response = client.newCall(request).execute()

                response.body?.string()?.let {
                    json.decodeFromString(CardDetailsResponse.serializer(), it)
                }

            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    suspend fun check3ds(email: String, cardId: String): ThreeDSResponse? {
        return withContext(Dispatchers.IO) {
            try {
                val jsonBody = """
                    {
                        "useremail": "$email",
                        "cardid": "$cardId"
                    }
                """.trimIndent()

                val requestBody =
                    jsonBody.toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url(ApiConfig.BASE_URL + "subusercheck3ds")
                    .post(requestBody)
                    .addHeader("publickey", ApiConfig.PUBLIC_KEY)
                    .addHeader("secretkey", ApiConfig.SECRET_KEY)
                    .addHeader("Content-Type", "application/json")
                    .build()

                val response = client.newCall(request).execute()

                response.body?.string()?.let {
                    json.decodeFromString(ThreeDSResponse.serializer(), it)
                }

            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    suspend fun approve3ds(email: String, cardId: String, eventId: String): okhttp3.Response? {
        return withContext(Dispatchers.IO) {
            try {
                val jsonBody = """
                    {
                        "useremail": "$email",
                        "cardid": "$cardId",
                        "eventId": "$eventId"
                    }
                """.trimIndent()

                val requestBody =
                    jsonBody.toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url(ApiConfig.BASE_URL + "subuserapprove3ds")
                    .post(requestBody)
                    .addHeader("publickey", ApiConfig.PUBLIC_KEY)
                    .addHeader("secretkey", ApiConfig.SECRET_KEY)
                    .addHeader("Content-Type", "application/json")
                    .build()

                client.newCall(request).execute()

            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    suspend fun blockDigitalCard(email: String, cardId: String): ApplyCardResponse? {
        return withContext(Dispatchers.IO) {
            try {
                val jsonBody = """
                {
                    "useremail": "$email",
                    "cardid": "$cardId"
                }
            """.trimIndent()

                val requestBody =
                    jsonBody.toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url(ApiConfig.BASE_URL + "subuserblockdigital")
                    .post(requestBody)
                    .addHeader("publickey", ApiConfig.PUBLIC_KEY)
                    .addHeader("secretkey", ApiConfig.SECRET_KEY)
                    .addHeader("Content-Type", "application/json")
                    .build()

                val response = client.newCall(request).execute()
                val responseBody = response.body?.string()
                Log.d("CardApiService", "blockDigitalCard response: $responseBody")

                responseBody?.let {
                    json.decodeFromString(ApplyCardResponse.serializer(), it)
                }

            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    suspend fun unblockDigitalCard(email: String, cardId: String): ApplyCardResponse? {
        return withContext(Dispatchers.IO) {
            try {
                val jsonBody = """
                {
                    "useremail": "$email",
                    "cardid": "$cardId"
                }
            """.trimIndent()

                val requestBody =
                    jsonBody.toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url(ApiConfig.BASE_URL + "subuserunblockdigital")
                    .post(requestBody)
                    .addHeader("publickey", ApiConfig.PUBLIC_KEY)
                    .addHeader("secretkey", ApiConfig.SECRET_KEY)
                    .addHeader("Content-Type", "application/json")
                    .build()

                val response = client.newCall(request).execute()
                val responseBody = response.body?.string()
                Log.d("CardApiService", "unblockDigitalCard response: $responseBody")

                responseBody?.let {
                    json.decodeFromString(ApplyCardResponse.serializer(), it)
                }

            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }
}

File: mastercard/lib/data/services/card_api_service.dart
import 'dart:convert';
import 'package:dio/dio.dart';
import '../../core/network/api_config.dart';
import '../models/apply_card_request.dart';
import '../models/apply_card_response.dart';
import '../models/card_details_response.dart';
import '../models/card_response.dart';
import '../models/sub_user.dart';
import '../models/three_ds_response.dart';

class CardApiService {
  static final Dio _dio = Dio(
    BaseOptions(
      baseUrl: ApiConfig.baseUrl,
      connectTimeout: const Duration(seconds: 15),
      receiveTimeout: const Duration(seconds: 30),
      headers: {
        'publickey': ApiConfig.publicKey,
        'secretkey': ApiConfig.secretKey,
        'Content-Type': 'application/json',
      },
    ),
  );

  static Future<ApplyCardResponse?> subuserAdd(SubUser subUser) async {
    try {
      final response = await _dio.post('subuseradd', data: subUser.toJson());
      if (response.data != null) {
        var responseData = response.data;
        if (responseData is String) {
          responseData = jsonDecode(responseData);
        }
        return ApplyCardResponse.fromJson(responseData);
      }
    } catch (e) {
      print("subuseradd error: $e");
    }
    return null;
  }

  static Future<ApplyCardResponse?> applyForNewVirtualCard(ApplyCardRequest request) async {
    try {
      final response = await _dio.post('digitalnewsubusercard', data: request.toJson());
      if (response.data != null) {
        var responseData = response.data;
        if (responseData is String) {
          responseData = jsonDecode(responseData);
        }
        return ApplyCardResponse.fromJson(responseData);
      }
    } catch (e) {
      print("applyForNewVirtualCard error: $e");
    }
    return null;
  }

  static Future<CardResponse?> getAllDigitalCards(String userEmail) async {
    try {
      final data = {'useremail': userEmail};
      print("getAllDigitalCards request: $data");
      final response = await _dio.post('getsubuseralldigital', data: data);
      
      var responseData = response.data;
      print("getAllDigitalCards response type: ${responseData.runtimeType}");
      
      if (responseData is String) {
        try {
          responseData = jsonDecode(responseData);
        } catch (e) {
          print("Error decoding JSON string: $e");
          return null;
        }
      }

      if (responseData != null && responseData is Map<String, dynamic>) {
        return CardResponse.fromJson(responseData);
      } else if (responseData != null) {
        print("Unexpected response data type: ${responseData.runtimeType}");
      }
    } catch (e) {
      print("getAllDigitalCards error: $e");
    }
    return null;
  }

  static Future<CardDetailsResponse?> getDigitalCardDetails(String email, String cardId) async {
    try {
      final response = await _dio.post('getsubuserdigitalcard', data: {
        'useremail': email,
        'cardid': cardId,
      });
      if (response.data != null) {
        var responseData = response.data;
        if (responseData is String) {
          responseData = jsonDecode(responseData);
        }
        return CardDetailsResponse.fromJson(responseData);
      }
    } catch (e) {
      print("getDigitalCardDetails error: $e");
    }
    return null;
  }

  static Future<ThreeDSResponse?> check3ds(String email, String cardId) async {
    try {
      final response = await _dio.post('subusercheck3ds', data: {
        'useremail': email,
        'cardid': cardId,
      });
      if (response.data != null) {
        var responseData = response.data;
        if (responseData is String) {
          responseData = jsonDecode(responseData);
        }
        return ThreeDSResponse.fromJson(responseData);
      }
    } on DioException catch (e) {
      print("check3ds DioException: ${e.response?.statusCode} - ${e.response?.data}");
      if (e.response?.statusCode == 422) {
        return ThreeDSResponse(status: "error", code: "422");
      }
      if (e.response?.data != null) {
        var responseData = e.response!.data;
        if (responseData is String) {
          try {
            responseData = jsonDecode(responseData);
          } catch (_) {}
        }
        if (responseData is Map<String, dynamic>) {
          return ThreeDSResponse.fromJson(responseData);
        }
      }
    } catch (e) {
      print("check3ds error: $e");
    }
    return null;
  }

  static Future<bool> approve3ds(String email, String cardId, String eventId) async {
    try {
      final response = await _dio.post('subuserapprove3ds', data: {
        'useremail': email,
        'cardid': cardId,
        'eventId': eventId,
      });
      return response.statusCode == 200;
    } catch (e) {
      print("approve3ds error: $e");
    }
    return false;
  }

  static Future<ApplyCardResponse?> blockDigitalCard(String email, String cardId) async {
    try {
      final response = await _dio.post('subuserblockdigital', data: {
        'useremail': email,
        'cardid': cardId,
      });
      if (response.data != null) {
        var responseData = response.data;
        if (responseData is String) {
          responseData = jsonDecode(responseData);
        }
        return ApplyCardResponse.fromJson(responseData);
      }
    } catch (e) {
      print("blockDigitalCard error: $e");
    }
    return null;
  }

  static Future<ApplyCardResponse?> unblockDigitalCard(String email, String cardId) async {
    try {
      final response = await _dio.post('subuserunblockdigital', data: {
        'useremail': email,
        'cardid': cardId,
      });
      if (response.data != null) {
        var responseData = response.data;
        if (responseData is String) {
          responseData = jsonDecode(responseData);
        }
        return ApplyCardResponse.fromJson(responseData);
      }
    } catch (e) {
      print("unblockDigitalCard error: $e");
    }
    return null;
  }
}
